### PR TITLE
Run code generation only if needed

### PR DESCRIPTION
**Pull request by bitmouse**
Code generator is now run only when interface file change or adding
is detected.

### CODE REVIEW

File: cmake/faceliftMacros.cmake [hunk 1]
```diff
@@ -136,67 +136,98 @@ function(facelift_generate_code )
     set(multiValueArgs IMPORT_FOLDERS)
     cmake_parse_arguments(ARGUMENT "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
 
+    # generate only interfaces which has changed since last build
+    file(GLOB_RECURSE QFACE_FILES ${ARGUMENT_INTERFACE_DEFINITION_FOLDER}/*.qface)
```
Comment (jacky309) [inline cmake/faceliftMacros.cmake]: you also need to check if the code generator itself has changed.

Comment (bitmouse) [reply cmake/faceliftMacros.cmake]: Fixed: line 167

File: cmake/faceliftMacros.cmake [hunk 2]
```diff
@@ -136,67 +136,103 @@ function(facelift_generate_code )
...
 
-    facelift_synchronize_folders(${WORK_PATH} ${ARGUMENT_OUTPUT_PATH})
+        facelift_synchronize_folders(${WORK_PATH} ${ARGUMENT_OUTPUT_PATH})
 
-    # Delete work folder
-    file(REMOVE_RECURSE ${WORK_PATH})
+        # Delete work folder
+        file(REMOVE_RECURSE ${WORK_PATH})
+    else()
+        message("No interface files modified. Code generation ignored.")
```
Comment (jacky309) [inline cmake/faceliftMacros.cmake]: "Code generation ignored" ??

File: cmake/faceliftMacros.cmake [hunk 3]
```diff
@@ -136,67 +136,103 @@ function(facelift_generate_code )
...
+
+    set(NUM_QFACE_FILES_${LIBRARY_NAME} ${NUM_QFACE_FILES} CACHE INTERNAL "Number of qface files for library")
+
     get_target_property(CODEGEN_EXECUTABLE_LOCATION facelift-codegen LOCATION)
     get_filename_component(CODEGEN_LOCATION ${CODEGEN_EXECUTABLE_LOCATION} DIRECTORY)
     set(QFACE_BASE_LOCATION ${CODEGEN_LOCATION}/qface)
 
-    file(TO_NATIVE_PATH "${QFACE_BASE_LOCATION}" QFACE_BASE_LOCATION_NATIVE_PATH)
-    set(ENV{PYTHONPATH} "${QFACE_BASE_LOCATION_NATIVE_PATH}")
+    file(TIMESTAMP ${CODEGEN_EXECUTABLE_LOCATION} FACELIFT_GENERATOR_TIMESTAMP "%s")
```
Comment (bitmouse) [inline cmake/faceliftMacros.cmake]: > Changing a file of the code generator itself does not trigger the code generation, which is wrong.

actually it does (tested)

File: cmake/faceliftMacros.cmake [hunk 4]
```diff
@@ -136,67 +136,103 @@ function(facelift_generate_code )
...
-    file(REMOVE_RECURSE ${WORK_PATH})
+        # Delete work folder
+        file(REMOVE_RECURSE ${WORK_PATH})
+    else()
+        message("No interface files modified for target ${LIBRARY_NAME}. Code generation not started.")
+    endif()
 
     # Add a dependency so that CMake will reconfigure whenever one of the interface files is changed, which will refresh our generated files
-    file(GLOB_RECURSE QFACE_FILES ${ARGUMENT_INTERFACE_DEFINITION_FOLDER}/*.qface)
     file(GLOB_RECURSE CODEGEN_FILES ${CODEGEN_LOCATION}/*)
```
Comment (jacky309) [inline cmake/faceliftMacros.cmake]: A change to any of these files should also cause the code generator to be executed.

File: cmake/faceliftMacros.cmake [hunk 5]
```diff
@@ -136,67 +136,106 @@ function(facelift_generate_code )
...
 
-    file(TO_NATIVE_PATH "${QFACE_BASE_LOCATION}" QFACE_BASE_LOCATION_NATIVE_PATH)
-    set(ENV{PYTHONPATH} "${QFACE_BASE_LOCATION_NATIVE_PATH}")
-
-    set(WORK_PATH ${CMAKE_CURRENT_BINARY_DIR}/facelift_generated_tmp)
-    file(MAKE_DIRECTORY ${WORK_PATH})
-
-    set(BASE_CODEGEN_COMMAND ${CODEGEN_EXECUTABLE_LOCATION} --input "${ARGUMENT_INTERFACE_DEFINITION_FOLDER}" --output "${WORK_PATH}")
+    set(MODIFIED_FILES)
+    foreach(QFACE_FILE ${QFACE_FILES})
```
Comment (jacky309) [inline cmake/faceliftMacros.cmake]: foreach(QFACE_FILE ${QFACE_FILES} ${CODEGEN_FILES})

File: cmake/faceliftMacros.cmake [hunk 6]
```diff
@@ -136,67 +136,106 @@ function(facelift_generate_code )
...
+    file(GLOB_RECURSE CODEGEN_FILES ${CODEGEN_LOCATION}/*)
 
-    file(TO_NATIVE_PATH "${QFACE_BASE_LOCATION}" QFACE_BASE_LOCATION_NATIVE_PATH)
-    set(ENV{PYTHONPATH} "${QFACE_BASE_LOCATION_NATIVE_PATH}")
-
-    set(WORK_PATH ${CMAKE_CURRENT_BINARY_DIR}/facelift_generated_tmp)
-    file(MAKE_DIRECTORY ${WORK_PATH})
-
-    set(BASE_CODEGEN_COMMAND ${CODEGEN_EXECUTABLE_LOCATION} --input "${ARGUMENT_INTERFACE_DEFINITION_FOLDER}" --output "${WORK_PATH}")
+    set(MODIFIED_FILES)
```
Comment (jacky309) [inline cmake/faceliftMacros.cmake]: unset

File: cmake/faceliftMacros.cmake [hunk 7]
```diff
@@ -136,67 +136,106 @@ function(facelift_generate_code )
...
+    foreach(QFACE_FILE ${QFACE_FILES})
+        file(TIMESTAMP ${QFACE_FILE} MOD_TIME "%s")
+        if(MOD_TIME GREATER FACELIFT_LAST_BUILD_TIMESTAMP_${LIBRARY_NAME})
+            list(APPEND MODIFIED_FILES ${QFACE_FILE})
+        endif()
+    endforeach()
 
-    foreach(IMPORT_FOLDER ${ARGUMENT_IMPORT_FOLDERS})
-        list(APPEND BASE_CODEGEN_COMMAND "--dependency" "${IMPORT_FOLDER}")
+    foreach(CODEGEN_FILE ${CODEGEN_FILES})
```
Comment (jacky309) [inline cmake/faceliftMacros.cmake]: remove block

File: cmake/faceliftMacros.cmake [hunk 8]
```diff
@@ -136,67 +136,99 @@ function(facelift_generate_code )
     set(multiValueArgs IMPORT_FOLDERS)
     cmake_parse_arguments(ARGUMENT "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
 
+    # generate only interfaces which has changed since last build
```
Comment (jacky309) [inline cmake/faceliftMacros.cmake]: "Call the code generator if either the interface or the code generator has changed"

File: cmake/faceliftMacros.cmake [hunk 9]
```diff
@@ -136,67 +136,98 @@ function(facelift_generate_code )
...
+        # Delete work folder
+        file(REMOVE_RECURSE ${WORK_PATH})
+    else()
+        message("No interface files modified for target ${LIBRARY_NAME}. Code generation not started.")
+    endif()
 
     # Add a dependency so that CMake will reconfigure whenever one of the interface files is changed, which will refresh our generated files
-    file(GLOB_RECURSE QFACE_FILES ${ARGUMENT_INTERFACE_DEFINITION_FOLDER}/*.qface)
-    file(GLOB_RECURSE CODEGEN_FILES ${CODEGEN_LOCATION}/*)
     set_property(DIRECTORY APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS "${CODEGEN_FILES};${QFACE_FILES}")
```
Comment (jacky309) [inline cmake/faceliftMacros.cmake]: set_property(DIRECTORY APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS "${CODEGEN_FILES};${QFACE_FILES};${ARGUMENT_INTERFACE_DEFINITION_FOLDER}")

**Pull request merged**